hazelcast/src/main/java/com/hazelcast/internal/util/ModularJavaUtils.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util;
+
+import static com.hazelcast.internal.util.ModularJavaUtils.PackageAccessRequirement.createRequirement;
+
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+
+/**
+ * Helper class for simplify work with Java module system (Java 9+) in older Java versions.
+ */
+public final class ModularJavaUtils {
+
+    private static final ILogger LOGGER = Logger.getLogger(ModularJavaUtils.class);
+
+    private ModularJavaUtils() {
+    }
+
+    public static String getHazelcastModuleName() {
+        if (!JavaVersion.isAtLeast(JavaVersion.JAVA_9)) {
+            return null;
+        }
+        try {
+            Method methodGetModule = Class.class.getMethod("getModule");
+            Class<?> classModule = Class.forName("java.lang.Module");
+            Method methodGetName = classModule.getMethod("getName");
+            Object moduleHazelcast = methodGetModule.invoke(Hazelcast.class);
+            return (String) methodGetName.invoke(moduleHazelcast);
+        } catch (Exception e) {
+            LOGGER.finest("Getting Hazelcast module name failed", e);
+            return null;
+        }
+    }
+
+    /**
+     * Prints warning to given {@link ILogger} if Hazelcast is not provided a sufficient access to Java internal packages on
+     * Java 9 and newer.
+     */
+    public static void checkJavaInternalAccess(ILogger logger) {
+        if (logger == null || !JavaVersion.isAtLeast(JavaVersion.JAVA_9)) {
```

Comment:
why `logger == null ||` ? how this can ever be true?

Reply:
If somebody calls the method with null parameter, we don't care about any package access problems, because we can't print them anywhere. The newly added constant `LOGGER` takes care only about printing exceptions when trace log level is enabled.